### ISSUE TITLE
RequestTimeout setting results in `getProtocolByName: does not exist (no such protocol name: tcp))`

### ISSUE THREAD

**Issue by crockeea** [snoyberg/http-client#374]
When sending several requests using `mapM_ (parseRequest ...)` to different destinations, I get many errors of the form
```
ghc: HttpExceptionRequest Request {
  host                 = "127.0.0.1"
  port                 = 45733
  secure               = True
  requestHeaders       = []
  path                 = "/"
  queryString          = ""
  method               = "POST"
  proxy                = Nothing
  rawBody              = False
  redirectCount        = 10
  responseTimeout      = ResponseTimeoutDefault
  requestVersion       = HTTP/1.1
}
 (InternalException (HostCannotConnect "127.0.0.1" [Network.Socket.socket: resource exhausted (Too many open files)]))
ghc: HttpExceptionRequest Request {
  host                 = "127.0.0.1"
  port                 = 40299
  secure               = True
  requestHeaders       = []
  path                 = "/"
  queryString          = ""
  method               = "POST"
  proxy                = Nothing
  rawBody              = False
  redirectCount        = 10
  responseTimeout      = ResponseTimeoutDefault
  requestVersion       = HTTP/1.1
}
 (ConnectionFailure Network.BSD.getProtocolByName: does not exist (no such protocol name: tcp))
```

This sounds like #307, but I'm only sending 25 requests, which hardly seems like DOS. Furthermore, I noticed that if I set  the Manager's managerResponseTimeout=responseTimeoutNone, I do not get any errors (everything works fine). After inspecting and playing with `getConnectionWrapper` in Network.HTTP.Client.Core, I see absolutely no reason why changing this setting affects this error. Yet it does, every single time.

There are several questions:
1. Why does changing the timeout setting affect whether I can call getProtocolbyName/run out of file descriptors?
2. I'm apparently getting *multiple* errors for each request I send (about 20 errors for each request, most of which are identical), which makes weeding through errors much harder. Why is that happening?
3. What is the root cause here?

**Comment by snoyberg** [comment]
I don't know why the timeouts are affecting things, but the root cause is
running out of fds.

**Comment by crockeea** [comment]
I would be very interested to learn why timeouts matter. I would like to enforce a timeout, but I'm unable to do so due to running out of file descriptors. It doesn't seem like I should be anywhere near the limit, so it seems like timeouts are somehow affecting the fds usage.

**Comment by domenkozar** [comment]
It seems like `network` module is calling https://hackage.haskell.org/package/network-bsd-2.8.1.0/docs/Network-BSD.html#v:getProtocolByName on each `connect` call, which seems quite wasteful.

**Comment by domenkozar** [comment]
Which is hidden behind `getProtocolNumber`, but that changed in network-3, so needs further investigation.

**Comment by domenkozar** [comment]
@crockeea I'm willing to dig into this if you give me code to reproduce it.

**Comment by domenkozar** [comment]
My conclusion is that network-3 has been refactored to avoid these lookups, so it should use fewer file handles.

I suggest upgrading to network-3 if you hit these issues. Sadly [it hasn't been bumped on stackage since May](https://github.com/commercialhaskell/stackage/issues/4528)

**Comment by domenkozar** [comment]
More about it here: https://blog.hercules-ci.com/2019/09/30/recent-cachix-downtime/

**Comment by swamp-agr** [comment]
Hi @crockeea @domenkozar leaking connections were fixed and `http-client-0.7.5` was released on Hackage: https://hackage.haskell.org/package/http-client

Please go ahead and re-test your cases!
